src/auth/src/credentials/mds_credential.rs

```
@@ -41,9 +44,74 @@ struct MDSCredential<T>
 where
     T: TokenProvider,
 {
+    endpoint: String,
+    quota_project_id: Option<String>,
+    universe_domain: RwLock<Option<String>>,
     token_provider: T,
 }
 
+#[bon]
+impl MDSCredential<MDSAccessTokenProvider> {
+    #[builder]
+    fn new(
+        scopes: Option<Vec<String>>,
+        quota_project_id: Option<String>,
+        universe_domain: Option<String>,
+        endpoint: String,
+    ) -> Self {
+        let token_provider = MDSAccessTokenProvider::builder()
+            .endpoint(endpoint.clone())
+            .maybe_scopes(scopes)
+            .build();
+
+        MDSCredential {
+            endpoint,
+            quota_project_id,
+            universe_domain: RwLock::new(universe_domain),
+            token_provider,
+        }
+    }
+}
+
+impl<T> MDSCredential<T>
+where
+    T: TokenProvider,
+{
+    async fn get_universe_domain_from_mds(&self) -> Result<String> {
```

We always ignore the error, should we consider returning `Option<String>` here?